crates/bevy_ecs/src/schedule/label.rs

```diff
@@ -0,0 +1,131 @@
+use std::{
+    any::Any,
+    borrow::Cow,
+    fmt::Debug,
+    hash::{Hash, Hasher},
+};
+
+use crate::{StageLabelMarker, SystemLabelMarker};
+
+pub trait Label<T>: DynHash + DynClone<T> + Send + Sync + 'static {
+    fn name(&self) -> Cow<'static, str>;
```

Comment:
Using a special `name` allows us to return a `Cow<'static, str>` without, which avoids an allocation when we need to access the name

Reply:
I agree, this is cleaner. However, it'd be amazing if there was some way to make the macro's default use `Debug`/`Display` somehow. (I'm not entirely sure what exactly does it do now, use the type name?)

Reply:
Currently, it uses the type name, but allows the user to inject an arbitrary expression with `#[name_expr]` so it can be overriden. I'm not sure if we want that API though, so its uncertain

Reply:
Thinking about this more, do we need to essentially cache the name at this level? As far as I can tell, this is only used for error reporting right now - all of which crash, because they cannot be recovered from. In the future this will be used by the editor, which will most likely cache its own strings, including those with labels in them. I think a `Debug` (and potentially Display) requirement on `Label<T>` is the right way to go.